core/store/src/trie/mem/flexible_data/encoding.rs

```diff
@@ -0,0 +1,137 @@
+use crate::trie::mem::arena::{Arena, ArenaPtr, ArenaPtrMut, ArenaSliceMut};
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::io::Write;
+
+use super::FlexibleDataHeader;
+
+/// Implementing this trait indicates that the type can be borsh serialized and
+/// the length of the serialization is a constant.
+///
+/// TODO: it would be nice to generate this, or test the correctness of the
+/// `SERIALIZED_SIZE` specification. For now, we rely on higher-level testing.
+pub trait BorshFixedSize {
+    const SERIALIZED_SIZE: usize;
+}
+
+/// Facilitates allocation and encoding of flexibly-sized data.
+pub struct RawEncoder<'a> {
+    data: ArenaSliceMut<'a>,
+    pos: usize,
+}
+
+// To make it easier to use borsh serialization.
+impl<'a> Write for RawEncoder<'a> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.data.raw_slice_mut()[self.pos..self.pos + buf.len()].copy_from_slice(buf);
+        self.pos += buf.len();
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+impl<'a> RawEncoder<'a> {
+    /// Creates a new arena allocation of the given size, returning an encoder
+    /// that can be used to initialize the allocated memory.
+    pub fn new(arena: &'a mut Arena, n: usize) -> RawEncoder<'a> {
+        let data = arena.alloc(n);
+        RawEncoder { data, pos: 0 }
+    }
+
+    /// Encodes the given fixed-size field to the current encoder position,
+    /// and then advances the position by the encoded size of the field.
+    pub fn encode<T: BorshSerialize>(&mut self, data: T) {
+        data.serialize(self).unwrap();
+    }
+
+    /// Encodes the given flexibly-sized part of the data to the current
+    /// encoder position, and then advances the position by the size of the
+    /// flexibly-sized part, as returned by `header.flexible_data_length()`.
```

Comment:
Also comment that the header itself is not being encoded by this function, and must be encoded independently by the caller